### ISSUE TITLE
Adam Optimizer Performance Improvements

### ISSUE THREAD

**Issue by chaseWillden** [delta-rs/delta#142]
According to the flame charts, the adam optimizer is very slow. We need to improve the performance

**Comment by mjovanc** [comment]
Absolutely, I added two benchmarks for the Adam optimizer that we can take advantage of (just need a slight improvement, check comments in code): https://github.com/delta-rs/delta/pull/148

**Comment by mjovanc** [comment]
Also we need to benchmark it using metal.

**Comment by mjovanc** [comment]
Test comment.
